ide/lsp.client/src/org/netbeans/modules/lsp/client/debugger/breakpoints/BreakpointAnnotationProvider.java

```diff
@@ -0,0 +1,270 @@
+import java.util.Set;
+
+import org.netbeans.api.debugger.Breakpoint;
+import org.netbeans.api.debugger.Breakpoint.VALIDITY;
+import org.netbeans.api.debugger.DebuggerManager;
+import org.netbeans.api.debugger.DebuggerManagerAdapter;
+
+import org.openide.filesystems.FileObject;
+import org.openide.loaders.DataObject;
+import org.openide.text.Annotation;
+import org.openide.text.AnnotationProvider;
+import org.openide.text.Line;
+import org.openide.util.Lookup;
+import org.openide.util.RequestProcessor;
+import org.openide.util.WeakListeners;
+import org.openide.util.WeakSet;
+
+
+/**
+ * This class is called when some file in editor is opened. It changes if
+ * some breakpoints are added or removed.
+ *
+ */
+@org.openide.util.lookup.ServiceProvider(service=org.openide.text.AnnotationProvider.class)
+public final class BreakpointAnnotationProvider extends DebuggerManagerAdapter implements AnnotationProvider {
+
+    private final Map<DAPLineBreakpoint, Set<Annotation>> breakpointToAnnotations = new IdentityHashMap<>();
+    private final Set<FileObject> annotatedFiles = new WeakSet<>();
+    private Set<PropertyChangeListener> dataObjectListeners;
+    private volatile boolean breakpointsActive = true;
+    private final RequestProcessor annotationProcessor = new RequestProcessor("CPP BP Annotation Refresh", 1);
+
+    public BreakpointAnnotationProvider() {
+        DebuggerManager.getDebuggerManager().addDebuggerListener(DebuggerManager.PROP_BREAKPOINTS, this);
+    }
+
+    @Override
+    public void annotate (Line.Set set, Lookup lookup) {
+        final FileObject fo = lookup.lookup(FileObject.class);
+        if (fo != null) {
+            DataObject dobj = lookup.lookup(DataObject.class);
+            if (dobj != null) {
+                PropertyChangeListener pchl = new PropertyChangeListener() {
+                    /** annotate renamed files. */
+                    @Override
+                    public void propertyChange(PropertyChangeEvent evt) {
+                        if (DataObject.PROP_PRIMARY_FILE.equals(evt.getPropertyName())) {
+                            DataObject dobj = (DataObject) evt.getSource();
+                            final FileObject newFO = dobj.getPrimaryFile();
+                            annotationProcessor.post(new Runnable() {
+                                @Override
+                                public void run() {
+                                    annotate(newFO);
+                                }
+                            });
+                        }
+                    }
+                };
+                dobj.addPropertyChangeListener(WeakListeners.propertyChange(pchl, dobj));
+                synchronized (this) {
+                    if (dataObjectListeners == null) {
+                        dataObjectListeners = new HashSet<>();
+                    }
+                    // Prevent from GC.
+                    dataObjectListeners.add(pchl);
+                }
```

Comment:
I think the comment here is wrong here and the usage of a `WeakListener` looks useless to me. The AnnotationProvider looks like a singleton, so this is essentially a memory leak as the listeners stay around indefinitely.

Reply:
Good catch. Fixed as of https://github.com/apache/netbeans/pull/8098#issuecomment-2571615395